go/db/tls.go

```diff
@@ -54,42 +58,100 @@ func init() {
+//   and adjust the settings as appropriate
+// the mysqlURI is the one _without_ TLS configuration settings
+func requiresTLS(host string, port int, mysqlURI string) bool {
+	const (
+		selectQuery = `
+			select
+				required
+			from
+				database_instance_tls
+			where
+				hostname = ?
+				and port = ?
+		`
+		insertOrUpdateQuery = `
+			insert into
+			database_instance_tls (
+				hostname, port, required
+			) values (
+				?, ?, ?
+			)
+			on duplicate key update
+			required=values(required)
+		`
+	)
+
+	var (
+		requiredInBackend      int
+		requiredWhenConnecting int
+		foundInBackend         bool
+		foundInCache           bool
+		value                  interface{}
+	)
 	cacheKey := fmt.Sprintf("%s:%d", host, port)
 
-	if value, found := requireTLSCache.Get(cacheKey); found {
+	// Check the value from the cache and return that if available
+	if value, foundInCache = requireTLSCache.Get(cacheKey); foundInCache {
 		readInstanceTLSCacheCounter.Inc(1)
-		return value.(bool)
-	}
 
-	required := false
-	db, _, _ := sqlutils.GetDB(mysql_uri)
-	if err := db.Ping(); err != nil && strings.Contains(err.Error(), Error3159) {
-		required = true
+		return value.(int) != 0
 	}
 
-	query := `
-			insert into
-				database_instance_tls (
-					hostname, port, required
-				) values (
-					?, ?, ?
-				)
-				on duplicate key update
-					required=values(required)
-				`
-	if _, err := ExecOrchestrator(query, host, port, required); err != nil {
+	// Check from the backend the expected value to use
+	// - note: this value may be wrong as the server configuration may have changed.
+	err := QueryOrchestrator(selectQuery, sqlutils.Args(host, port), func(m sqlutils.RowMap) error {
+		foundInBackend = true
+		requiredInBackend = m.GetInt("required")
+		return nil
+	})
+	readInstanceTLSCounter.Inc(1)
+	if err != nil {
 		log.Errore(err)
+		return requiredInBackend != 0 // this value may be wrong!
+	}
+
+	// Try to connect to the instance using the normal uri
+	db, _, _ := sqlutils.GetDB(mysqlURI)
+	err = db.Ping()
+	if err != nil {
+		// TLS required message
+		if strings.Contains(err.Error(), Error3159) {
+			requiredWhenConnecting = 1
+		} else if strings.Contains(err.Error(), ErrorTLSNotSupported) {
+			requiredWhenConnecting = 0
+		}
 	}
-	writeInstanceTLSCounter.Inc(1)
 
-	requireTLSCache.Set(cacheKey, required, cache.DefaultExpiration)
+	// update the cache
+	requireTLSCache.Set(fmt.Sprintf("%s:%d", host, port), requiredWhenConnecting, cache.DefaultExpiration)
 	writeInstanceTLSCacheCounter.Inc(1)
 
-	return required
+	// figure out if we need to update the backend database
+	if !foundInBackend || requiredInBackend != requiredWhenConnecting {
+		_, err = ExecOrchestrator(insertOrUpdateQuery, host, port, requiredWhenConnecting)
+		if err != nil {
+			log.Errore(err)
+			return requiredWhenConnecting != 0
+		}
+
+		writeInstanceTLSCounter.Inc(1)
+	}
+
+	return requiredWhenConnecting != 0
```

Comment:
What's the reason for `requiredWhenConnecting` being an `int` rather than `bool`?

Reply:
The column in the database is a `tinyint unsigned` (setup that way previously) so I'm not bothering to change it but to use the value retrieved etc.

```
root@my-orchestrator-backend [orchestrator]> show create table database_instance_tls\G
*************************** 1. row ***************************
       Table: database_instance_tls
Create Table: CREATE TABLE `database_instance_tls` (
  `hostname` varchar(128) NOT NULL,
  `port` smallint(5) unsigned NOT NULL,
  `required` tinyint(3) unsigned NOT NULL DEFAULT '0',
  PRIMARY KEY (`hostname`,`port`)
) ENGINE=InnoDB DEFAULT CHARSET=ascii
1 row in set (0.00 sec)
```

Reply:
You can have `requiredWhenConnecting` as a `bool` and it will transparently translate to/from an `INT` in the database.

Reply:
@sjmudd would you like to change to `bool`?